libraries/standard/mqtt/src/mqtt.c

```diff
@@ -99,6 +101,509 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+
+        if( bytesReceived != ( int32_t ) bytesToReceive )
+        {
+            IotLogErrorWithArgs( "Receive error while discarding packet."
+                                 "ReceivedBytes=%d, ExpectedBytes=%u.",
+                                 bytesReceived,
+                                 bytesToReceive );
+            shouldBreak = true;
+        }
+        else
+        {
+            totalBytesReceived += ( uint32_t ) bytesReceived;
+            /* Update remaining time and check for timeout. */
+            elapsedTime = calculateElapsedTime( getTimeStamp(), entryTime );
+            if( elapsedTime < timeoutMs )
+            {
+                remainingTimeMs = timeoutMs - elapsedTime;
+            }
+            else
+            {
+                IotLogError( "Time expired while discarding packet." );
+                shouldBreak = true;
+            }
+        }
+
+        if( shouldBreak )
+        {
+            break;
+        }
+    }
+
+    if( totalBytesReceived == remainingLength )
+    {
+        IotLogErrorWithArgs( "Dumped packet. DumpedBytes=%d.",
+                             totalBytesReceived );
+        /* Packet dumped, so no data is available. */
+        status = MQTTNoDataAvailable;
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Receive a packet from the transport interface.
+ *
+ * @param[in] pContext MQTT Connection context.
+ * @param[in] incomingPacket packet struct with remaining length.
+ * @param[in] remainingTimeMs Time remaining to receive the packet.
+ *
+ * @return #MQTTSuccess or #MQTTRecvFailed.
+ */
+static MQTTStatus_t receivePacket( MQTTContext_t * const pContext,
+                                   MQTTPacketInfo_t incomingPacket,
+                                   uint32_t remainingTimeMs )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    int32_t bytesReceived = 0;
+    size_t bytesToReceive = 0U;
+
+    assert( pContext != NULL );
+
+    if( incomingPacket.remainingLength > pContext->networkBuffer.size )
+    {
+        IotLogErrorWithArgs( "Incoming packet length %u exceeds network buffer size %u."
+                             "Incoming packet will be dumped.",
+                             incomingPacket.remainingLength
+                             pContext->networkBuffer );
+        status = discardPacket( pContext,
+                                incomingPacket.remainingLength,
+                                remainingTimeMs );
+    }
+    else
+    {
+        bytesToReceive = incomingPacket.remainingLength;
+        bytesReceived = recvExact( pContext, bytesToReceive, remainingTimeMs );
+        if( bytesReceived == ( int32_t ) bytesToReceive )
+        {
+            /* Receive successful, bytesReceived == bytesToReceive. */
+            IotLogInfoWithArgs( "Packet received. ReceivedBytes=%d.",
+                                bytesReceived );
```

Comment:
I think we would need to point the buffer read to MQTTPacketInfo.pRemainingData

Reply:
https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/901/files#diff-34cd787ed56bd43628017def1ba1bae0R1064

Reply:
Noticed that this is already handled in MQTT_ProcessLoop.